velox/substrait/VeloxToSubstraitType.cpp

```diff
@@ -0,0 +1,336 @@
+      return *sType;
+    }
+    case velox::TypeKind::INTEGER: {
+      sType->set_allocated_i32(new ::substrait::Type_I32());
+      return *sType;
+    }
+    case velox::TypeKind::BIGINT: {
+      sType->set_allocated_i64(new ::substrait::Type_I64());
+      return *sType;
+    }
+    case velox::TypeKind::REAL: {
+      sType->set_allocated_fp32(new ::substrait::Type_FP32());
+      return *sType;
+    }
+    case velox::TypeKind::DOUBLE: {
+      sType->set_allocated_fp64(new ::substrait::Type_FP64());
+      return *sType;
+    }
+    case velox::TypeKind::VARCHAR: {
+      sType->set_allocated_varchar(new ::substrait::Type_VarChar());
+      return *sType;
+    }
+    case velox::TypeKind::VARBINARY: {
+      sType->set_allocated_binary(new ::substrait::Type_Binary());
+      return *sType;
+    }
+    case velox::TypeKind::TIMESTAMP: {
+      sType->set_allocated_timestamp(new ::substrait::Type_Timestamp());
+      return *sType;
+    }
+    case velox::TypeKind::ARRAY: {
+      ::substrait::Type_List* sTList = new ::substrait::Type_List();
+      const std::shared_ptr<const Type> vArrayType =
+          vType->asArray().elementType();
+      ::substrait::Type sListType =
+          toSubstraitType(vArrayType, sTList->mutable_type());
+
+      sType->set_allocated_list(sTList);
+      return *sType;
+    }
+    case velox::TypeKind::MAP: {
+      ::substrait::Type_Map* sMap = new ::substrait::Type_Map();
+      const std::shared_ptr<const Type> vMapKeyType = vType->asMap().keyType();
+      const std::shared_ptr<const Type> vMapValueType =
+          vType->asMap().valueType();
+
+      toSubstraitType(vMapKeyType, sMap->mutable_key());
+      toSubstraitType(vMapValueType, sMap->mutable_value());
+
+      sType->set_allocated_map(sMap);
+      return *sType;
+    }
+    case velox::TypeKind::UNKNOWN:
+    case velox::TypeKind::FUNCTION:
+    case velox::TypeKind::OPAQUE:
+    case velox::TypeKind::INVALID:
+    default:
+      throw std::runtime_error(
+          "Unsupported type " + std::string(vType->kindName()));
+  }
+}
+
+::substrait::NamedStruct* VeloxToSubstraitTypeConvertor::toSubstraitNamedStruct(
+    const velox::RowTypePtr& vRow,
+    ::substrait::NamedStruct* sNamedStruct) {
+  int64_t vSize = vRow->size();
+  std::vector<std::string> vNames = vRow->names();
+  std::vector<std::shared_ptr<const Type>> vTypes = vRow->children();
+
+  for (int64_t i = 0; i < vSize; ++i) {
+    std::string vName = vNames.at(i);
+    std::shared_ptr<const Type> vType = vTypes.at(i);
+    sNamedStruct->add_names(vName);
+    ::substrait::Type* sStruct = sNamedStruct->mutable_struct_()->add_types();
+
+    toSubstraitType(vType, sStruct);
+  }
+
+  return sNamedStruct;
+}
+
+::substrait::Expression_Literal*
+VeloxToSubstraitTypeConvertor::processVeloxValueByType(
+    ::substrait::Expression_Literal_Struct* sLitValue,
+    ::substrait::Expression_Literal* sField,
+    const velox::VectorPtr& children) {
```

Comment:
The name "children" here is strange. Should it be "vector"?

Reply:
change it.